x/stake/simulation/sim_test.go

```diff
@@ -0,0 +1,369 @@
+		denom := k.GetParams(ctx).BondDenom
+		validatorKey := keys[r.Intn(len(keys))]
+		validatorAddress := sdk.AccAddress(validatorKey.PubKey().Address())
+		delegatorKey := keys[r.Intn(len(keys))]
+		delegatorAddress := sdk.AccAddress(delegatorKey.PubKey().Address())
+		amount := m.GetAccount(ctx, delegatorAddress).GetCoins().AmountOf(denom)
+		if amount.GT(sdk.ZeroInt()) {
+			amount = sdk.NewInt(int64(r.Intn(int(amount.Int64()))))
+		}
+		if amount.Equal(sdk.ZeroInt()) {
+			return "nop", nil
+		}
+		msg := stake.MsgBeginUnbonding{
+			DelegatorAddr: delegatorAddress,
+			ValidatorAddr: validatorAddress,
+			SharesAmount:  sdk.NewRatFromInt(amount),
+		}
+		require.Nil(t, msg.ValidateBasic(), "expected msg to pass ValidateBasic: %s", msg.GetSignBytes())
+		ctx, write := ctx.CacheContext()
+		result := stake.NewHandler(k)(ctx, msg)
+		if result.IsOK() {
+			write()
+		}
+		stats[fmt.Sprintf("stake/beginunbonding/%v", result.IsOK())] += 1
+		action = fmt.Sprintf("TestMsgBeginUnbonding: %s", msg.GetSignBytes())
+		return action, nil
+	}
+}
+
+// SimulateMsgCompleteUnbonding
+func SimulateMsgCompleteUnbonding(k stake.Keeper) mock.TestAndRunMsg {
+	return func(t *testing.T, r *rand.Rand, ctx sdk.Context, keys []crypto.PrivKey, log string) (action string, err sdk.Error) {
+		validatorKey := keys[r.Intn(len(keys))]
+		validatorAddress := sdk.AccAddress(validatorKey.PubKey().Address())
+		delegatorKey := keys[r.Intn(len(keys))]
+		delegatorAddress := sdk.AccAddress(delegatorKey.PubKey().Address())
+		msg := stake.MsgCompleteUnbonding{
+			DelegatorAddr: delegatorAddress,
+			ValidatorAddr: validatorAddress,
+		}
+		require.Nil(t, msg.ValidateBasic(), "expected msg to pass ValidateBasic: %s", msg.GetSignBytes())
+		ctx, write := ctx.CacheContext()
+		result := stake.NewHandler(k)(ctx, msg)
+		if result.IsOK() {
+			write()
+		}
+		stats[fmt.Sprintf("stake/completeunbonding/%v", result.IsOK())] += 1
+		action = fmt.Sprintf("TestMsgCompleteUnbonding with %s", msg.GetSignBytes())
+		return action, nil
+	}
+}
+
+// SimulateMsgBeginRedelegate
+func SimulateMsgBeginRedelegate(m auth.AccountMapper, k stake.Keeper) mock.TestAndRunMsg {
+	return func(t *testing.T, r *rand.Rand, ctx sdk.Context, keys []crypto.PrivKey, log string) (action string, err sdk.Error) {
+		denom := k.GetParams(ctx).BondDenom
+		sourceValidatorKey := keys[r.Intn(len(keys))]
+		sourceValidatorAddress := sdk.AccAddress(sourceValidatorKey.PubKey().Address())
+		destValidatorKey := keys[r.Intn(len(keys))]
+		destValidatorAddress := sdk.AccAddress(destValidatorKey.PubKey().Address())
+		delegatorKey := keys[r.Intn(len(keys))]
+		delegatorAddress := sdk.AccAddress(delegatorKey.PubKey().Address())
+		// TODO
+		amount := m.GetAccount(ctx, delegatorAddress).GetCoins().AmountOf(denom)
+		if amount.GT(sdk.ZeroInt()) {
+			amount = sdk.NewInt(int64(r.Intn(int(amount.Int64()))))
+		}
+		if amount.Equal(sdk.ZeroInt()) {
+			return "nop", nil
```

Comment:
what is the "nop" action? "no-operation"? let's document all this clearly (if it's not already)

Reply:
Changed to `no-operation`.